powershell-adapter/powershell.resource.ps1

```diff
@@ -1,427 +1,349 @@
-
-    exit 1
+# module types
+enum moduleType {
+    ScriptBased
+    ClassBased
 }
 
-Import-Module $DscModule -DisableNameChecking
-
-# Adding some debug info to STDERR
-$m = gmo PSDesiredStateConfiguration
-$trace = @{"Debug"="PSVersion="+$PSVersionTable.PSVersion.ToString()} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="PSPath="+$PSHome} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="ModuleVersion="+$m.Version.ToString()} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="ModulePath="+$m.Path} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-$trace = @{"Debug"="PSModulePath="+$env:PSModulePath} | ConvertTo-Json -Compress
-$host.ui.WriteErrorLine($trace)
-
-if ($Operation -eq 'List')
-{
-    $DscResources= Get-DscResource
-
-    foreach ($r in $DscResources)
-    {
-        if ($r.ImplementedAs -eq "Binary")
-        {
+# Cache the results of Get-DscResource to optimize performance
+function Invoke-CacheRefresh {
+    # cache the results of Get-DscResource
+    [resourceCache[]]$resourceCache = @()
+    $DscResources = Get-DscResource
+    foreach ($dsc in $DscResources) {
+        # only support known moduleType, excluding binary
+        if ([moduleType].GetEnumNames() -notcontains $dsc.ImplementationDetail) {
             continue
         }
-
-        $version_string = "";
-        if ($r.Version) { $version_string = $r.Version.ToString() }
-        $author_string = "";
-        if ($r.author) { $author_string = $r.CompanyName.ToString() }
-        $moduleName = "";
-        if ($r.ModuleName) { $moduleName = $r.ModuleName }
-        elseif ($r.ParentPath) { $moduleName = Split-Path $r.ParentPath | Split-Path | Split-Path -Leaf }
-
-        $propertyList = @()
-        foreach ($p in $r.Properties)
-        {
-            if ($p.Name)
-            {
-                $propertyList += $p.Name
-            }
+        # workaround: if the resource does not have a module name, get it from parent path
+        # workaround: modulename is not settable, so clone the object without being read-only
+        $DscResourceInfo = [dscResource]::new()
+        $dsc.PSObject.Properties | ForEach-Object -Process { $DscResourceInfo.$($_.Name) = $_.Value }
+        if ($dsc.ModuleName) {
+            $moduleName = $dsc.ModuleName
         }
-
-        $fullResourceTypeName = "$moduleName/$($r.ResourceType)"
-        $script:ResourceCache[$fullResourceTypeName] = $r
-        if ($WinPS) {$requiresString = "Microsoft.Windows/WindowsPowerShell"} else {$requiresString = "Microsoft.DSC/PowerShell"}
-
-        $t = [Type]$r.ResourceType
-        $exportMethod = $t.GetMethod('Export')
-
-        $capabilities = @('Get', 'Set', 'Test')
-        if ($null -ne $exportMethod) {
-            $capabilities += 'Export'
+        elseif ($dsc.ParentPath) {
+            $moduleName = Split-Path $dsc.ParentPath | Split-Path | Split-Path -Leaf
```

Comment:
Can you explain why `split-path` is needed 3 times here?

Reply:
this carried over from previous work, but I believe that is the proper number of steps to get back to the module name folder

Reply:
I see.  Please add a comment in the code explaining this.

Reply:
added